File: PetMedicalCare/BackEnd/SoftPetBusiness/src/main/java/pe/edu/pucp/softpet/bo/VeterinarioBo.java
package pe.edu.pucp.softpet.bo;

import java.sql.Date;
import java.util.ArrayList;
import pe.edu.pucp.softpet.bo.utils.GmailService;
import pe.edu.pucp.softpet.daoImp.VeterinarioDaoImpl;
import pe.edu.pucp.softpet.bo.utils.SecurityUtil;
import pe.edu.pucp.softpet.dto.personas.PersonaDto;
import pe.edu.pucp.softpet.dto.personas.VeterinarioDto;
import pe.edu.pucp.softpet.dto.util.enums.EstadoVeterinario;

public class VeterinarioBo {

    private final VeterinarioDaoImpl veterinarioDao;
    
    private final GmailService gmailService;

    public VeterinarioBo() {
        this.veterinarioDao = new VeterinarioDaoImpl();
        
        this.gmailService = new GmailService(); 
    }

    public Integer insertar(int personaId, String fechaContratacion,
            String estado, String especialidad, boolean activo) {

        VeterinarioDto veterinario = new VeterinarioDto();

        PersonaDto persona = new PersonaDto();
        persona.setPersonaId(personaId);

        veterinario.setPersona(persona);
        veterinario.setFechaContratacion(Date.valueOf(fechaContratacion));
        veterinario.setEstado(EstadoVeterinario.valueOf(estado.toUpperCase()));
        veterinario.setEspecialidad(especialidad);
        veterinario.setActivo(activo);

        return this.veterinarioDao.insertar(veterinario);
    }

    public Integer modificar(int veterinarioId, int personaId, String fechaContratacion,
            String estado, String especialidad, boolean activo) {

        VeterinarioDto veterinario = new VeterinarioDto();

        PersonaDto persona = new PersonaDto();
        persona.setPersonaId(personaId);

        veterinario.setVeterinarioId(veterinarioId);
        veterinario.setPersona(persona);
        veterinario.setFechaContratacion(Date.valueOf(fechaContratacion));
        veterinario.setEstado(EstadoVeterinario.valueOf(estado.toUpperCase()));
        veterinario.setEspecialidad(especialidad);
        veterinario.setActivo(activo);

        return this.veterinarioDao.modificar(veterinario);
    }

    public Integer eliminar(int veterinarioId) {
        VeterinarioDto veterinario = new VeterinarioDto();
        veterinario.setVeterinarioId(veterinarioId);
        return this.veterinarioDao.eliminar(veterinario);
    }

    public VeterinarioDto obtenerPorId(int veterinarioId) {
        return this.veterinarioDao.obtenerPorId(veterinarioId);
    }

    public ArrayList<VeterinarioDto> listarTodos() {
        return this.veterinarioDao.listarTodos();
    }

    public ArrayList<VeterinarioDto> listarVeterinariosActivos() {
        return this.veterinarioDao.listarVeterinariosActivos();
    }

    public int VerificarSiExisteHorarioLaboral(String fecha, Integer idVeterinario) {

        java.sql.Date FechaDate = java.sql.Date.valueOf(fecha);
        Integer resultado = veterinarioDao.VerificarSiExisteHorarioLaboral(FechaDate, idVeterinario);
        return resultado;

    }

    public ArrayList<VeterinarioDto> ListasBusquedaAvanzadaVeterinario(
            String Especialidad,
            String nombre,
            String nroDocumento,
            Integer estadoActivo) { // Parametro INT
        return (ArrayList<VeterinarioDto>) this.veterinarioDao.ListasBusquedaAvanzadaVeterinario(
                Especialidad, nombre, nroDocumento, estadoActivo);
    }

//    public Integer insertarVeterinarioCompleto(
//            String username, String password, String correo, boolean activoUsuario,
//            String nombre, String direccion, String telefono, String sexo,
//            Integer nroDocumento, Integer ruc, String tipoDocumento,
//            String fechaContratacion, String estado, String especialidad) {
//        return this.veterinarioDao.insertarVeterinarioCompleto(
//                username, password, correo, activoUsuario,
//                nombre, direccion, telefono, sexo,
//                nroDocumento, ruc, tipoDocumento,
//                fechaContratacion, estado, especialidad);
//    }
//    
//    // ... (Métodos existentes) ...
//
//    // NUEVO: Modificar Completo
//    public Integer modificarVeterinarioCompleto(
//            Integer idVeterinario, Integer idPersona, Integer idUsuario,
//            String username, String password, String correo, boolean activo,
//            String nombre, String direccion, String telefono, String sexo, 
//            Integer nroDocumento, Integer ruc, String tipoDocumento,
//            String fechaContratacion, String estado, String especialidad) {
//        
//        return this.veterinarioDao.modificarVeterinarioCompleto(
//                idVeterinario, idPersona, idUsuario,
//                username, password, correo, activo,
//                nombre, direccion, telefono, sexo,
//                nroDocumento, ruc, tipoDocumento,
//                fechaContratacion, estado, especialidad);
//    }
    
    // MÉTODO CLAVE: Insertar Completo con Seguridad
    public Integer insertarVeterinarioCompleto(
            String username, String passwordIgnorada, String correo, boolean activoUsuario,
            String nombre, String direccion, String telefono, String sexo,
            Integer nroDocumento, Integer ruc, String tipoDocumento,
            String fechaContratacion, String estado, String especialidad) {
        
        // 1. GENERAR CONTRASEÑA ALEATORIA (Ej: "Xy9Za2")
        String passwordRaw = SecurityUtil.generarPasswordAleatoria();
        
        // 2. ENCRIPTARLA (HASH) PARA LA BD
        String passwordHashed = SecurityUtil.sha256(passwordRaw);
        
        // 3. ENVIAR CORREO CON LA CONTRASEÑA ORIGINAL (RAW)
        // Usamos un hilo aparte para que el guardado no sea lento
        new Thread(() -> {
            this.gmailService.enviarCorreo_Credenciales(correo, passwordRaw);
        }).start();

        // 4. GUARDAR EN BD USANDO EL HASH
        // Llamamos al DAO pasándole 'passwordHashed' en lugar de la original
        return this.veterinarioDao.insertarVeterinarioCompleto(
                username, passwordHashed, correo, activoUsuario,
                nombre, direccion, telefono, sexo,
                nroDocumento, ruc, tipoDocumento,
                fechaContratacion, estado, especialidad);
    }
    
    // ... (Mantén modificarCompleto y eliminarCompleto igual, la modificación de password es aparte) ...
    
    public Integer modificarVeterinarioCompleto(
            Integer idVeterinario, Integer idPersona, Integer idUsuario,
            String username, String password, String correo, boolean activo,
            String nombre, String direccion, String telefono, String sexo, 
            Integer nroDocumento, Integer ruc, String tipoDocumento,
            String fechaContratacion, String estado, String especialidad) {
        
        // Si viene password, podríamos encriptarlo aquí también, 
        // pero generalmente la secretaria NO cambia la contraseña, lo hace el usuario.
        // Por seguridad, si password viene vacío o null, el DAO lo ignora.
        return this.veterinarioDao.modificarVeterinarioCompleto(
                idVeterinario, idPersona, idUsuario,
                username, password, correo, activo,
                nombre, direccion, telefono, sexo,
                nroDocumento, ruc, tipoDocumento,
                fechaContratacion, estado, especialidad);
    }

    // NUEVO: Eliminar Completo
    public Integer eliminarVeterinarioCompleto(Integer idVeterinario) {
        return this.veterinarioDao.eliminarVeterinarioCompleto(idVeterinario);
    }
}

File: Mis últimos pasos/SoftPetWA/SoftPetBussiness/SOAPBo/GestionNOLogistica/VeterinarioBo.cs
using SoftPetBussiness.ServicioClient;
using SoftPetBussiness.VeterinarioClient;
using System;
using System.Collections.Generic;
using System.Linq;

namespace SoftPetBusiness
{
    public class VeterinarioBO
    {
        private VeterinariosClient clienteSOAP;

        public VeterinarioBO()
        {
            this.clienteSOAP = new VeterinariosClient();
        }

        // Insertar veterinario (Parámetros sueltos)
        public int Insertar(int personaId, string fechaContratacion, string estado, string especialidad, bool activo)
        {
            return this.clienteSOAP.insertar_veterinario(personaId, fechaContratacion, estado, especialidad, activo);
        }

        // Modificar veterinario (Parámetros sueltos)
        public int Modificar(int veterinarioId, int personaId, string fechaContratacion, string estado, string especialidad, bool activo)
        {
            return this.clienteSOAP.modificar_veterinario(veterinarioId, personaId, fechaContratacion, estado, especialidad, activo);
        }

        // Eliminar veterinario
        public int Eliminar(int veterinarioId)
        {
            return this.clienteSOAP.eliminar_veterinario(veterinarioId);
        }

        // Obtener veterinario por ID
        public veterinarioDto ObtenerPorId(int veterinarioId)
        {
            return this.clienteSOAP.obtenerPorId_veterinario(veterinarioId);
        }

        // Listar todos los veterinarios
        public List<veterinarioDto> ListarTodos()
        {
            return this.clienteSOAP.listar_veterinarios().ToList<veterinarioDto>();
        }

        public List<veterinarioDto> ListarActivos()
        {
            return this.clienteSOAP.listar_veterinarios_activos().ToList<veterinarioDto>();
        }

        // Insertar veterinario usando DTO (FIX: Formato de fecha)
        public int Insertar(veterinarioDto veterinario)
        {
            return this.clienteSOAP.insertar_veterinario(
                veterinario.persona.personaId,
                veterinario.fechaContratacion.ToString("yyyy-MM-dd"), // FIX AQUÍ
                veterinario.estado.ToString(),
                veterinario.especialidad,
                veterinario.activo
            );
        }

        // Modificar veterinario usando DTO (FIX: Formato de fecha)
        public int Modificar(veterinarioDto veterinario)
        {
            return this.clienteSOAP.modificar_veterinario(
                veterinario.veterinarioId,
                veterinario.persona.personaId,
                veterinario.fechaContratacion.ToString("yyyy-MM-dd"), // FIX AQUÍ
                veterinario.estado.ToString(),
                veterinario.especialidad,
                veterinario.activo
            );
        }

        public List<veterinarioDto> ListarBusquedaAvanzada(string nombre, string documento, string especialidad, int estadoActivo)
        {
            nombre = nombre ?? "";
            documento = documento ?? "";
            especialidad = (especialidad == "Todos") ? "" : (especialidad ?? "");

            // Llamada al WS con el entero
            var resultado = this.clienteSOAP.ListasBusquedaAvanzadaVeterinario(especialidad, nombre, documento, estadoActivo);

            if (resultado == null) return new List<veterinarioDto>();
            return resultado.ToList<veterinarioDto>();
        }

        public int InsertarCompleto(
            string username, string password, string correo,
            string nombre, string direccion, string telefono, string sexo,
            int nroDocumento, int ruc, string tipoDocumento,
            string fechaContratacion, string estado, string especialidad)
        {
            return this.clienteSOAP.insertarVeterinarioCompleto(
                username, password, correo,
                nombre, direccion, telefono, sexo,
                nroDocumento, ruc, tipoDocumento,
                fechaContratacion, estado, especialidad
            );
        }

        public int ModificarCompleto(
            int idVeterinario, int idPersona, int idUsuario,
            string username, string password, string correo, bool activo,
            string nombre, string direccion, string telefono, string sexo,
            int nroDocumento, int ruc, string tipoDocumento,
            string fechaContratacion, string estado, string especialidad)
        {
            // Llamada al método 'modificarVeterinarioCompleto' del WS Java
            return this.clienteSOAP.modificarVeterinarioCompleto(
                idVeterinario, idPersona, idUsuario,
                username, password, correo, activo,
                nombre, direccion, telefono, sexo,
                nroDocumento, ruc, tipoDocumento,
                fechaContratacion, estado, especialidad
            );
        }

        public int EliminarCompleto(int idVeterinario)
        {
            // Llamada al método 'eliminarVeterinarioCompleto' del WS Java
            return this.clienteSOAP.eliminarVeterinarioCompleto(idVeterinario);
        }
    }
}